pkgs/applications/editors/neovim/wrapper.nix

```diff
@@ -55,44 +55,89 @@ let
+          else
+            # Tells nvim to practically not load this provider
+            "'let g:loaded_${prog}_provider=1'"
+          )
+        + "\"")
+      ]
+    ) hostprog_check_table))
+    ++ lib.optionals withRuby [
+      "--set" "GEM_HOME" "${rubyEnv}/${rubyEnv.ruby.gemPath}"
+    ]
+    ++ lib.optionals (binPath != "") [
+      "--suffix" "PATH" binPath
+    ]
+    # this relies on a patched neovim, see
+    # https://github.com/neovim/neovim/issues/9413
+    ++ lib.optionals (configure != {}) [
+      "--set" "NVIM_SYSTEM_RPLUGIN_MANIFEST" "$out/rplugin.vim"
+      "--add-flags" "-u" "${vimUtils.vimrcFile configure}"
+    ]
+  ;
+
   in
-  stdenv.mkDerivation {
+  symlinkJoin {
       name = "neovim-${stdenv.lib.getVersion neovim}";
-      buildCommand = let bin="${neovim}/bin/nvim"; in ''
-        if [ ! -x "${bin}" ]
-        then
-            echo "cannot find executable file \`${bin}'"
-            exit 1
-        fi
-
-        makeWrapper "$(readlink -v --canonicalize-existing "${bin}")" \
-          "$out/bin/nvim" --add-flags " \
-        --cmd \"${if withNodeJs then "let g:node_host_prog='${nodePackages.neovim}/bin/neovim-node-host'" else "let g:loaded_node_provider=1"}\" \
-        --cmd \"${if withPython then "let g:python_host_prog='$out/bin/nvim-python'" else "let g:loaded_python_provider = 1"}\" \
-        --cmd \"${if withPython3 then "let g:python3_host_prog='$out/bin/nvim-python3'" else "let g:loaded_python3_provider = 1"}\" \
-        --cmd \"${if withRuby then "let g:ruby_host_prog='$out/bin/nvim-ruby'" else "let g:loaded_ruby_provider=1"}\" " \
-        --suffix PATH : ${binPath} \
-        ${optionalString withRuby '' --set GEM_HOME ${rubyEnv}/${rubyEnv.ruby.gemPath}'' }
-      ''
-      + optionalString (!stdenv.isDarwin) ''
-        # copy icon and patch the original neovim.desktop file
-        mkdir -p $out/share/{applications,pixmaps}
-        ln -s ${neovim}/share/pixmaps/nvim.png $out/share/pixmaps/nvim.png
+      postBuild = ''
+        # Remove the symlinks created by symlinkJoin which we need to perform
+        # extra actions upon
+        rm $out/share/applications/nvim.desktop $out/bin/nvim
+        makeWrapper ${builtins.concatStringsSep " " makeWrapperArgs} ${extraMakeWrapperArgs}
         substitute ${neovim}/share/applications/nvim.desktop $out/share/applications/nvim.desktop \
           --replace 'TryExec=nvim' "TryExec=$out/bin/nvim" \
           --replace 'Name=Neovim' 'Name=WrappedNeovim'
       ''
       + optionalString withPython ''
         makeWrapper ${pythonEnv}/bin/python $out/bin/nvim-python --unset PYTHONPATH
-      '' + optionalString withPython3 ''
+      ''
+      + optionalString withPython3 ''
         makeWrapper ${python3Env}/bin/python3 $out/bin/nvim-python3 --unset PYTHONPATH
-      '' + optionalString withRuby ''
+      ''
+      + optionalString withRuby ''
         ln -s ${rubyEnv}/bin/neovim-ruby-host $out/bin/nvim-ruby
```

Comment:
I wonder why we need theses symlinks ? we should just set the providers directly to their interpreter.

Reply:
@teto note this line:

https://github.com/doronbehar/nixpkgs/blob/42d6fc768605e059fb39b7ecb00539f800109b63/pkgs/applications/editors/neovim/wrapper.nix#L88

Part of the point of the PR, is to **treat all providers equally** - so the same Nix code would be used to tell the final `makeWrapper` call to use any of those providers. Because we do need to further wrap the python interpreters, (and we don't need that for the ruby interpreters), I decided it'd be best to "expose" these providers via at `$out/bin/nvim-${provider}`. This way they'd be accessible also to non nix configurations for users that might want to refer to e.g `/run/current-system/sw/bin/nvim-python`.

The alternative, would be to introduce another table such as:

https://github.com/doronbehar/nixpkgs/blob/42d6fc768605e059fb39b7ecb00539f800109b63/pkgs/applications/editors/neovim/wrapper.nix#L69-L74

Which will hold the paths to each provider and these will be used in

https://github.com/doronbehar/nixpkgs/blob/42d6fc768605e059fb39b7ecb00539f800109b63/pkgs/applications/editors/neovim/wrapper.nix#L88

I tend to think it'll make the code much more complex for little to no value.